optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,330 @@
+import sys
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+
+import numpy as np
+
+import optuna
+from optuna import logging
+from optuna._experimental import experimental
+from optuna._imports import _LazyImport
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.distributions import CategoricalDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+_SUGGESTED_STATES = (TrialState.COMPLETE, TrialState.PRUNED)
+
+
+@experimental("3.0.0")
+class QMCSampler(BaseSampler):
+    """A Quasi Monte Carlo Sampler that generates low-discrepancy sequences.
+
+    Quasi Monte Carlo (QMC) sequences are designed to have lower discrepancies than
+    standard random seqeunces. They are known to perform better than the standard
+    randam sequences in hyperparameter optimization.
+
+    For further information about the use of QMC sequences for hyperparameter optimization,
+    please refer to the following paper:
+
+    - `Bergstra, James, and Yoshua Bengio. Random search for hyper-parameter optimization.
+      Journal of machine learning research 13.2, 2012.
+      <https://jmlr.org/papers/v13/bergstra12a.html>`_
+
+    We use the QMC implementations in Scipy. For the details of the QMC algorithm,
+    see the Scipy API references on `scipy.stats.qmc
+    <https://scipy.github.io/devdocs/reference/stats.qmc.html>`_.
+
+    .. note:
+        If your search space contains categorical parameters, it samples the catagorical
+        parameters by its `independent_sampler` without using QMC algorithm.
+
+    .. note::
+        The search space of the sampler is determined by either previous trials in the study or
+        the first trial that this sampler samples.
+
+        If there are previous trials in the study, :class:`~optuna.samplers.QMCSamper` infers its
+        search space using the trial which was created first in the study.
+
+        Otherwise (if the study has no previous trials), :class:`~optuna.samplers.QMCSampler`
+        samples the first trial using its `independent_sampler` and then infers the search space
+        in the second trial.
```

Comment:
Why is that? Shouldn't this use QMC instead? I don't know the details of Optuna, but could this explain the low performance in the benchmark if the sampler is not used to get the first search space?

Reply:
The QMC sequences are used from the second trial and later. We define the number of parameters and (their distributions) by sampling the first trial. Thus, how we sample parameters in the first trial actually does not matter at all.

Reply:
But the quality of your distribution estimation depends on the quality of the sample no? When I read all this, what I understand is: for a problem in 10D, generate an initial population of ex. 100, then compute some things, generate new points using QMC as the algorithm goes. If that's correct, then the initial sample could be as important as the rest as you need to properly discover the space to properly estimate what you want to compute.

Reply:
Ah, not actually. Even if the problem has 10 dimensions, we sample one parameter for each of 10 parameters in the first trial and then we evaluate the performance of the parameters (usually by the cross validation or test set performance.) The second trials and later is sampled using QMC sample. So if we can run cross-validation for 100 configurations (of 10 parameters), only 1 configuration is taken randomly and the other 99 configurations are taken from QMC. Let me know if this makes sense!

Reply:
Ok so a configuration is a sample of size `n`. And you repeat this let say 100 times. The first one is sampling `n` randomly while the rest uses QMC.

Just not sure I get the `10 dimensions, we sample one parameter for each of 10 parameters`. I read this as: take one dimension at a time and sample `n` values. And you use this to construct `(n, 100)`. Why is this done like that and not just sampling `(n, d)`? Or do you intentionally fix some params in the process to make some kind of sensitivity analysis? Is this what you mean with `evaluate the performance of the parameters`?

Reply:
What I tried to explain was to take (1, d) sample in the first trial using a random number generator and then sample (n-1, d) using QMC, when we need to optimize a d-dimensional function and we can afford to evaluate the function n times. So we are not interested in fixing some parameters in the first trial.